components/tidb_query/src/codec/mysql/json/binary.rs

```diff
@@ -54,52 +99,58 @@ impl Json {
         }
     }
 
+    /// Returns the encoding binary length of self
     pub fn binary_len(&self) -> usize {
-        TYPE_LEN + self.body_binary_len()
-    }
-
-    fn body_binary_len(&self) -> usize {
-        match *self {
-            Json::Object(ref d) => get_obj_binary_len(d),
-            Json::Array(ref d) => get_array_binary_len(d),
-            Json::Boolean(_) | Json::None => LITERAL_LEN,
-            Json::I64(_) | Json::U64(_) | Json::Double(_) => NUMBER_LEN,
-            Json::String(ref d) => get_str_binary_len(d),
-        }
-    }
-
-    fn get_type_code(&self) -> u8 {
-        match *self {
-            Json::Object(_) => TYPE_CODE_OBJECT,
-            Json::Array(_) => TYPE_CODE_ARRAY,
-            Json::Boolean(_) | Json::None => TYPE_CODE_LITERAL,
-            Json::I64(_) => TYPE_CODE_I64,
-            Json::U64(_) => TYPE_CODE_U64,
-            Json::Double(_) => TYPE_CODE_DOUBLE,
-            Json::String(_) => TYPE_CODE_STRING,
-        }
+        TYPE_LEN + self.value.len()
     }
 }
 
 pub trait JsonEncoder: NumberEncoder {
-    fn write_json(&mut self, data: &Json) -> Result<()> {
-        self.write_u8(data.get_type_code())?;
-        self.write_json_body(data)
+    fn write_json<'a>(&mut self, data: JsonRef<'a>) -> Result<()> {
+        self.write_u8(data.get_type() as u8)?;
+        self.write_bytes(data.value()).map_err(Error::from)
     }
 
-    fn write_json_body(&mut self, data: &Json) -> Result<()> {
-        match *data {
-            Json::Object(ref d) => self.write_json_obj(d),
-            Json::Array(ref d) => self.write_json_array(d),
-            Json::Boolean(_) | Json::None => {
-                let v = data.as_literal()?;
-                self.write_json_literal(v)
-            }
-            Json::I64(d) => self.write_json_i64(d),
-            Json::U64(d) => self.write_json_u64(d),
-            Json::Double(d) => self.write_json_f64(d),
-            Json::String(ref d) => self.write_json_str(d),
+    fn write_json_obj_from_keys_values<'a>(
```

Comment:
It's better to use map here to avoid the length of keys and values inconsistent.

Reply:
See: https://github.com/tikv/tikv/pull/6388/files#r367199199, we should use a generic method to deduplicate code.